File: app/src/main/java/com/example/ruslanyussupov/popularmovies/Result.java
package com.example.ruslanyussupov.popularmovies;


import android.support.annotation.Nullable;

public class Result<T> {

    @Nullable
    public final T data;

    @Nullable
    public final String error;

    public final State state;


    private Result(@Nullable T data, State state, @Nullable String error) {
        this.data = data;
        this.state = state;
        this.error = error;
    }

    public static <T> Result<T> success(T data) {
        return new Result<>(data, State.SUCCESS, null);
    }

    public static <T> Result<T> loading() {
        return new Result<>(null, State.LOADING, null);
    }

    public static <T> Result<T> error(String error) {
        return new Result<>(null, State.ERROR, error);
    }

    public enum State {
        LOADING, SUCCESS, ERROR
    }

}

File: app/src/main/java/com/example/ruslanyussupov/popularmovies/Result.kt
package com.example.ruslanyussupov.popularmovies

class Result<T> private constructor(val data: T?,
                                    val state: State,
                                    val error: String) {

    enum class State {
        LOADING, SUCCESS, ERROR
    }

    companion object {

        fun <T> success(data: T): Result<T> {
            return Result(data, State.SUCCESS, "")
        }

        fun <T> loading(): Result<T> {
            return Result(null, State.LOADING, "")
        }

        fun <T> error(error: String): Result<T> {
            return Result(null, State.ERROR, error)
        }
    }

}
